simulation/halsim_print/src/halsim_print/cpp/main.cpp

```diff
@@ -0,0 +1,29 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2016-2017 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include <iostream>
+
+#include "HALSimPrint.h"
+#include "PrintPWM.h"
+
+/* Currently, robots never terminate, so we keep a single static object
+   and it is never properly released or cleaned up. */
+static HALSimPrint halsim;
+
+extern "C" {
+#if defined(WIN32) || defined(_WIN32)
+__declspec(dllexport)
```

Comment:
Maybe not for this PR, but in the future we should add an extra header file (something like HALSim.h) to include shortcuts for these. Something like [here](https://github.com/JacisNonsense/ToastCPP/blob/master/Toast-Core/include/toast/library.hpp). `CAPI` is used for the init function.

`API` will have to be applied to all C++ symbols we want to call from the dlopen'd HAL sym, unless we choose to export symbols as a companion `.def` file on windows.

Reply:
We don't need DLLExports added explicitly. We handle this through a custom plugin in gradle, so all symbols are exported every build.

Reply:
All I know is that if I do not include that dllexport line, then the resulting DLL does not have the symbols required for the GetProcAddress to succeed.  If I need to flag this dll somehow in gradle, I'd appreciate a pointer on how to do that.

Reply:
Add the following to the model area for halsim_print. When you build, you will get some linker errors for missing symbols. For all the missing symbols, copy them, and add them to the ExcludeSymbols array. The x86 and x64 will most likely be different, so make sure you use the arch you are building. By default, gradle will not build both x86 and x64, so use `-PbuildAll` to cause gradle to build both x86 and x64 instead of just the current os.

```
exportsConfigs {
    halsim_print(ExportsConfig) {
        x86ExcludeSymbols = [ ]
        x64ExcludeSymbols = [ ]
    }
}
```

Reply:
Is that really appropriate in this case?  This dll is intended to export exactly one symbol.  Isn't it simpler to just put the declspec in the file?

Reply:
Given the single symbol export, I think I like the declspec better too, although it needs to be made cross-platform.

Reply:
There is no need for the declspec on Linux; non static functions are exported by default.  I believe macOS is the same way.

Reply:
So are we both 
- Never planning on linking to this library statically 
- Never planning on explicitly linking to the dll, or explicitly using the header?

If no to both questions, the explicit symbol export is fine, and we should disable the static build, and make the headers non public. If the answer is no to either of those question, we have to use the exportsConfig, otherwise those cases will break (statically needs no dllimport or dllexport), and linking to the dll needs either dllimport or nothing, and dllexport won't work. And adding a compiler flag to switch between the options is not easily doable, and not detectable by the compiler

Reply:
Yes, this library should never be used statically, nor did it publish any headers or static libraries.  I've pushed an update to the new PR that also prevents it from building the static library to hopefully make that more plain.